src/main/java/htsjdk/samtools/read/Read.java

```diff
@@ -0,0 +1,1438 @@
+     *
+     * @param flag flag to be set or unset.
+     * @param set {@code true} if the flag should be set; {@code false} otherwise.
+     *
+     * @throws IllegalArgumentException if the flag is {@code null}.
+     */
+    default void addFlag(final SAMFlag flag, final boolean set) {
+        if (flag == null) {
+            throw new IllegalArgumentException("null flag");
+        }
+        int bitwiseFlag = getFlag();
+        if (set) {
+            bitwiseFlag |= flag.intValue();
+        } else {
+            bitwiseFlag &= ~flag.intValue();
+        }
+        setFlag(bitwiseFlag);
+    }
+
+    /**
+     * Sets the bit of the provided flag.
+     *
+     * <p>Equivalent to <b>FLAG</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>Default implementation calls {@code addFlag(flag, true)}.
+     *
+     * @param flag flag to be set.
+     *
+     * @throws IllegalArgumentException if the flag is {@code null}.
+     */
+    default void setFlag(final SAMFlag flag) {
+        addFlag(flag, true);
+    }
+
+    /**
+     * Unsets the bit of the provided flag.
+     *
+     * <p>Equivalent to <b>FLAG</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>Default implementation calls {@code addFlag(flag, false)}.
+     *
+     * @param flag flag to be unset.
+     */
+    default void unsetFlag(final SAMFlag flag) {
+        addFlag(flag, false);
+    }
+
+    /**
+     * Sets all the provided flags.
+     *
+     * <p>Equivalent to <b>FLAG</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>Default implementation calls {@link #setFlag(SAMFlag)} for each of the elements in the set.
+     *
+     * @param flags flags to be set.
+     */
+    default void setFlags(final Set<SAMFlag> flags) {
+        flags.forEach(this::setFlag);
+    }
+
+    /**
+     * Unsets all the provided flags.
+     *
+     * <p>Equivalent to <b>FLAG</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>Default implementation calls {@link #unsetFlag(SAMFlag)}} for each of the elements in the set.
+     *
+     * @param flags flags to be unset.
+     */
+    default void unsetFlags(final Set<SAMFlag> flags) {
+        flags.forEach(this::unsetFlag);
+    }
+
+    /**
+     * Gets the contig name where the read is mapped to. May return null if there is no unique mapping.
+     *
+     * <p>Equivalent to <b>RNAME</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * @return name of the contig this read is mapped to; {@link ReadConstants#NO_ALIGNMENT_REFERENCE_NAME} if unset (e.g. for unmapped reads).
+     */
+    @Override
+    String getContig();
+
+    /**
+     * Gets the 1-based inclusive leftmost position of the sequence remaining after clipping.
```

Comment:
I would stick to the language from the spec: 

>1-based leftmost mapping POSition of the first CIGAR operation that “consumes” a reference base

in particular this doesn't include insertions, N operations and P operators (in addition to padding)

Reply:
Good point! That's a much better description...

Reply:
Done.